### PR TITLE
Lazy initialize Strategy.parallel_devices

### LINKED ISSUE #11097: Lazy initialize properties in Strategies
## Proposed refactor
Strategy has four properties in init : Accelerator, Precision_plugin, Checkpoint_io and Cluster_enviroment.

User could pass in a strategy class or a strategy str into trainer.
- If user pass a string, accelerator_connector will initialize the strategy with the properties selection. 
- If user pass a strategy class, the logics get a bit mess. We should track what user has passed in V.S strategy default value V.S. Trainer.accelerator_connector 

This issue propose lazy initialization for strategy properties. @awaelchli and @ananthsub also bought this up before

### Motivation
For correctness, maintenance and enable future simplifications 

### Pitch

Current Training_type_plugin __init__ will set default value to precision_plugin and checkpoint_io:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/training_type_plugin.py#L40-L58
```python
    def __init__(
        self, checkpoint_io: Optional[CheckpointIO] = None, precision_plugin: Optional[PrecisionPlugin] = None
    ) -> None:
        self._model: Optional[Module] = None
        checkpoint_io = checkpoint_io if checkpoint_io is not None else TorchCheckpointIO()
        self._checkpoint_io = checkpoint_io
        self._precision_plugin = precision_plugin if precision_plugin is not None else PrecisionPlugin()

```

Proposal

```python
def __init__(
        self, checkpoint_io: Optional[CheckpointIO] = None, precision_plugin: Optional[PrecisionPlugin] = None
    ) -> None:
        self._model: Optional[Module] = None
        self._checkpoint_io = checkpoint_io
        self._precision_plugin = precision_plugin

@property
def checkpoint_io(self):
       return self._checkpoint_io if self._checkpoint_io is not None else TorchCheckpointIO()

@setter
def checkpoint_io(checkpoint_io):
       self._checkpoint_io = checkpoint_io

@property
def precision_plugin(self):
      return self._precision_plugin if self._precision_plugin is not None else PrecisionPlugin()

@setter
def precision_plugin(precision_plugin):
       self._precision_plugin = precision_plugin
```

In accelerator_connector.py, we know:
`training_type_plugin._precision_plugin` before we call setter is what user passed in, not the default value.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @kaushikb11 @ananthsub

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?

Part of #11449
This is will simplify accelerator_connector refactor

Fixes #11097

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/strategies/parallel.py [hunk 1]
```diff
@@ -72,6 +72,14 @@ def world_size(self) -> int:
     def is_global_zero(self) -> bool:
         return self.global_rank == 0
 
+    @property
+    def parallel_devices(self):
+        return self._parallel_devices
```
Comment (carmocca) [inline pytorch_lightning/strategies/parallel.py:77]: Did you forget to rename the attribute in the `__init__`? L43

Comment (four4fish) [reply pytorch_lightning/strategies/parallel.py:77]: No, it's same style as 'accelerator' and etc properties. 
Like https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/strategies/strategy.py#L44-L85

I also feel this way of lazy initialization is confusing. Should we change all of them to 
```
def __init__(accelerator):
      self._accelerator = accelerator
     
@property 
def accelerator(self):
      return self._accelerator
```
?

Comment (carmocca) [reply pytorch_lightning/strategies/parallel.py:77]: Duh!

I think it's fine as it is.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py [hunk 2]
```diff
@@ -135,6 +134,10 @@ def num_nodes(self, num_nodes: int) -> None:
         self._num_nodes = num_nodes
         self.set_world_ranks()
 
+    @property
+    def num_processes(self):
+        return len(self.parallel_devices) if self.parallel_devices is not None else 0
```
Comment (rohitgr7) [inline pytorch_lightning/strategies/ddp.py:139]: there are some places in accelerator_connector where this is set manually? is that required?
for eg L898.

File: pytorch_lightning/strategies/ddp.py [hunk 3]
```diff
@@ -135,6 +134,10 @@ def num_nodes(self, num_nodes: int) -> None:
         self._num_nodes = num_nodes
         self.set_world_ranks()
 
+    @property
+    def num_processes(self):
```
Comment (rohitgr7) [inline pytorch_lightning/strategies/ddp.py:138]: will it work for ddp2? since ddp2 inherits DDP strategy, and num_processes is num_nodes in that case set manually in L900 of `accelerator_connector`.

File: pytorch_lightning/strategies/ddp_spawn.py [hunk 4]
```diff
@@ -107,6 +106,10 @@ def local_rank(self) -> int:
     def root_device(self):
         return self.parallel_devices[self.local_rank]
 
+    @property
+    def num_processes(self):
```
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/strategies/ddp_spawn.py:110]: The previous version has self.num_processes set in __init__(), so TPUSpawnStrategy has `num_processes` property too.

File: pytorch_lightning/strategies/ddp.py (hunk 3 above)
Comment (four4fish) [reply pytorch_lightning/strategies/ddp.py:138]: accelerator_connector has self.num_processes too, but it's not equal to strategy.num_processes. They are not necessarily the same value. 

Only place accelerator_connector set strategy.num_processes is the L762-L763 before(modified below). 

The` accelerator_connector.num_processes` could be num_nodes, `parallel_devices` or `num_processes`, which equal to `trainer.num_processes` and been used in logging and profilings. 

The `Strategies.num_processes` is calculated from `strategies.parallel_devices`, which is unrelated to `accelerator_connector.num_proceses`. accelerator_connector need to set it manually base on "parallel_devices" in L762-L763 is because the `strategies.parallel_devices` is not lazy initialized

This is a pre-step for #11448, we should definitely revisit all the device related logics after #11448

File: pytorch_lightning/strategies/ddp.py (hunk 2 above)
Comment (four4fish) [reply pytorch_lightning/strategies/ddp.py:139]: same as above

File: pytorch_lightning/strategies/ddp.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/strategies/ddp.py:138]: ah! you are right... got a little confused with the same name.
Thanks for clarifying :)

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py (hunk 3 above)
Comment (four4fish) [reply pytorch_lightning/strategies/ddp.py:138]: Yeah, it's really confusing, and I think the devices related logic probably is wrong somewhere too. We can revisit them after the accelerator_connector rewrite

File: CHANGELOG.md [hunk 5]
```diff
@@ -222,6 +222,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed `MisconfigurationException` to `ModuleNotFoundError` when `rich` isn't available ([#11360](https://github.com/PyTorchLightning/pytorch-lightning/pull/11360))
 
 
+- Changed `parallel_devices` property in `ParallelStrategy` to be lazy initialized ([#](https://github.com/PyTorchLightning/pytorch-lightning/pull/))
+
+
```
Comment (akihironitta) [inline CHANGELOG.md:227]:
```suggestion
- Changed `parallel_devices` property in `ParallelStrategy` to be lazy initialized ([#11572](https://github.com/PyTorchLightning/pytorch-lightning/pull/11572))


```

Comment (four4fish) [reply CHANGELOG.md]: Thank you @akihironitta !!!

File: pytorch_lightning/strategies/parallel.py [hunk 6]
```diff
@@ -72,6 +72,14 @@ def world_size(self) -> int:
...
+
+    @parallel_devices.setter
+    def parallel_devices(self, parallel_devices):
+        self._parallel_devices = parallel_devices
```
Comment (awaelchli) [inline pytorch_lightning/strategies/parallel.py:81]: here we see a hint at the difference between the proposals #11097 and #7650. 

If a setter like this one gets introduced, and other values depend on this value here, these would have to be re-computed everytime this setter is invoked (depending on the requirements of the plugin, error handling would be needed to). If multiple of these setters are necessary, things get complicated and logic gets duplicated. #7650 would solve this, because it would set all attributes in one shot together. I haven't found the time to investigate in this direction further.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/strategies/parallel.py (hunk 6 above)
Comment (four4fish) [reply pytorch_lightning/strategies/parallel.py:81]: +1 for #7650. The accelerator_connector should be the only place allowed to set/modify these parameters (other than to pass in when creating the object). With #7650 proposal, we can call the lazy_init at the end of the accelerator_connector __init__, which will be way less error-prone and the behavior for properties will be more consistent and robust

This PR take the  #11097 easy way for now to unblock accelerator_connector rewrite..

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]
